In [2]:
import pandas as pd
from pandas_datareader import data

 - Series - 1D Data structure
 - Dataframe - 2D Data structure
 - Panel - 3D Data structure
   - A container of dataframes
   - Need to know 3 things: Data frame in panel to get data from, points in the dataframe to get specific value
 

In [4]:
companies = ["MSFT", "GOOG", "AAPL", "YHOO", "AMZN"]
p = data.DataReader(companies, data_source = 'google', start = '2010-01-01', end='2016-12-31')
p 
# This returns a panel object

<class 'pandas.core.panel.Panel'>
Dimensions: 5 (items) x 1745 (major_axis) x 5 (minor_axis)
Items axis: Open to Volume
Major_axis axis: 2010-01-04 00:00:00 to 2016-12-07 00:00:00
Minor_axis axis: AAPL to YHOO

# Section 11; Part 147
The axes of a panel object

 - `Dimensions` - Tells how many dataframes are in the panel (5 in panel above). It is **NOT** each of the 5 tech companies In each dataframe there are 1745 rows and 5 columns in each (the colums, in this case the companies). 
 - `Items Axis` - Outer layer of panel
 - `Major Axis` - The rows in each dataframe
 - `Minor Axis` - The columns of each dataframe (in this case, the companies)
 - `axes` - All of the axes in a single result

In [5]:
p.items

Index([u'Open', u'High', u'Low', u'Close', u'Volume'], dtype='object')

In [6]:
p.major_axis

DatetimeIndex(['2010-01-04', '2010-01-05', '2010-01-06', '2010-01-07',
               '2010-01-08', '2010-01-11', '2010-01-12', '2010-01-13',
               '2010-01-14', '2010-01-15',
               ...
               '2016-11-23', '2016-11-25', '2016-11-28', '2016-11-29',
               '2016-11-30', '2016-12-01', '2016-12-02', '2016-12-05',
               '2016-12-06', '2016-12-07'],
              dtype='datetime64[ns]', name=u'Date', length=1745, freq=None)

In [7]:
p.minor_axis

Index([u'AAPL', u'AMZN', u'GOOG', u'MSFT', u'YHOO'], dtype='object')

In [8]:
p.axes

[Index([u'Open', u'High', u'Low', u'Close', u'Volume'], dtype='object'),
 DatetimeIndex(['2010-01-04', '2010-01-05', '2010-01-06', '2010-01-07',
                '2010-01-08', '2010-01-11', '2010-01-12', '2010-01-13',
                '2010-01-14', '2010-01-15',
                ...
                '2016-11-23', '2016-11-25', '2016-11-28', '2016-11-29',
                '2016-11-30', '2016-12-01', '2016-12-02', '2016-12-05',
                '2016-12-06', '2016-12-07'],
               dtype='datetime64[ns]', name=u'Date', length=1745, freq=None),
 Index([u'AAPL', u'AMZN', u'GOOG', u'MSFT', u'YHOO'], dtype='object')]

# Section 11; Part 148
Panel Attributes

In [10]:
# ndim - Number of dimensions (panel is 3)
p.ndim

3

In [11]:
# dtypes - Series of index labels of each data frame (the .items attribute)
p.dtypes

Open      float64
High      float64
Low       float64
Close     float64
Volume    float64
dtype: object

In [13]:
# shape - Tuple of the measurements of the panel
# (number of dataframe, number of rows, number of columns)
p.shape

(5, 1745, 5)

In [14]:
# size - Total number of values in the panel
p.size

43625

In [15]:
# Can prove this value by multiplying values in .shape tuple
p.shape[0] * p.shape[1] * p.shape[2]

43625

In [16]:
# values - Array of values
#  Nested structure
p.values

array([[[  3.04900000e+01,   1.36250000e+02,   3.13160000e+02,
           3.06200000e+01,   1.69400000e+01],
        [  3.06600000e+01,   1.33430000e+02,   3.13280000e+02,
           3.08500000e+01,   1.72200000e+01],
        [  3.06300000e+01,   1.34600000e+02,   3.12620000e+02,
           3.08800000e+01,   1.71700000e+01],
        ..., 
        [  1.10000000e+02,   7.45000000e+02,   7.57710000e+02,
           5.97000000e+01,   4.00200000e+01],
        [  1.09500000e+02,   7.63990000e+02,   7.64730000e+02,
           6.04300000e+01,   4.03100000e+01],
        [  1.09260000e+02,   7.64550000e+02,   7.61000000e+02,
           6.00100000e+01,   3.99800000e+01]],

       [[  3.06400000e+01,   1.36610000e+02,   3.14440000e+02,
           3.11000000e+01,   1.72000000e+01],
        [  3.08000000e+01,   1.35480000e+02,   3.13610000e+02,
           3.11000000e+01,   1.72300000e+01],
        [  3.07500000e+01,   1.34730000e+02,   3.12620000e+02,
           3.10800000e+01,   1.73000000e+01],
   

# Section 11; Part 149
Use bracket notation to extract a dataframe (for subsetting)

 - Brackets are used to extract the next smallest datastructure
   - In Series, it extracts a value
   - In Dataframe, it extracts a column (which is a Series)
   - In Panel, it extracts a dataframe
 - To extract a dataframe, need to pass an item in the items axis via the bracket notation

In [17]:
# Get available values on the items axis
p.items

Index([u'Open', u'High', u'Low', u'Close', u'Volume'], dtype='object')

In [18]:
# Pull the Open dataframe
p['Open']

AAPL    AMZN    GOOG   MSFT   YHOO
Date                                            
2010-01-04   30.49  136.25  313.16  30.62  16.94
2010-01-05   30.66  133.43  313.28  30.85  17.22
2010-01-06   30.63  134.60  312.62  30.88  17.17
2010-01-07   30.25  132.01  304.40  30.63  16.81
2010-01-08   30.04  130.56  295.70  30.28  16.68
2010-01-11   30.40  132.62  301.93  30.71  16.77
2010-01-12   29.88  128.99  298.74  30.15  16.65
2010-01-13   29.70  127.90  287.96  30.26  16.88
2010-01-14   30.02  129.14  291.66  30.31  16.81
2010-01-15   30.13  129.18  295.50  31.08  17.25
2010-01-19   29.76  126.20  290.45  30.75  16.78
2010-01-20   30.70  127.13  292.70  30.81  16.65
2010-01-21   30.30  127.26  291.43  30.61  16.39
2010-01-22   29.54  125.60  281.97  30.00  16.08
2010-01-25   28.93  122.10  273.02  29.24  16.07
2010-01-26   29.42  120.56  268.72  29.20  15.82
2010-01-27   29.55  121.03  270.36  29.35  16.46
2010-01-28   29.28  124.43  271.97  29.84  15.93
2010-01-29   28.73  129.77  268.98  29.90  15.51
2010-02-01   27.48  123.18  267.03  28.39  15.14
2010-02-02   27.99  118.79  267.21  28.37  15.10
2010-02-03   27.88  117.12  264.07  28.26  15.12
2010-02-04   28.10  118.64  268.23  28.38  15.34
2010-02-05   27.52  115.88  263.94  28.00  15.01
2010-02-08   27.96  119.38  265.98  28.01  15.18
2010-02-09   28.06  118.20  269.50  27.97  15.20
2010-02-10   27.98  118.00  266.77  28.03  15.02
2010-02-11   27.84  117.21  265.86  27.93  14.87
2010-02-12   28.30  118.99  266.22  27.81  15.07
2010-02-16   28.85  120.06  268.30  28.13  15.23
...            ...     ...     ...    ...    ...
2016-10-26  114.31  832.76  806.34  60.81  42.26
2016-10-27  115.39  831.24  801.00  60.61  42.31
2016-10-28  113.87  782.00  808.35  60.01  42.03
2016-10-31  113.65  781.03  795.47  60.16  42.03
2016-11-01  113.46  799.00  782.89  59.97  41.13
2016-11-02  111.40  783.93  778.20  59.82  41.80
2016-11-03  110.98  765.05  767.25  59.53  40.62
2016-11-04  108.53  762.79  750.66  58.65  39.96
2016-11-07  110.08  771.64  774.50  59.78  41.17
2016-11-08  110.31  784.97  783.40  60.55  40.97
2016-11-09  109.88  764.00  779.94  60.00  40.44
2016-11-10  111.09  778.81  791.17  60.48  41.55
2016-11-11  107.12  735.73  756.54  58.23  39.84
2016-11-14  107.71  745.51  755.60  59.02  40.36
2016-11-15  106.57  730.00  746.97  58.33  39.61
2016-11-16  106.70  739.88  755.20  58.94  40.10
2016-11-17  109.81  749.32  766.92  60.41  41.34
2016-11-18  109.72  761.00  771.37  60.78  41.48
2016-11-21  110.12  766.00  762.61  60.50  41.44
2016-11-22  111.95  788.17  772.63  60.98  41.20
2016-11-23  111.36  781.73  767.73  61.01  40.91
2016-11-25  111.47  786.50  764.26  60.30  41.08
2016-11-28  111.43  776.99  760.00  60.34  40.85
2016-11-29  110.78  768.00  771.53  60.65  41.43
2016-11-30  111.60  762.00  770.07  60.86  41.62
2016-12-01  110.36  752.41  757.44  60.11  41.00
2016-12-02  109.17  743.40  744.59  59.08  39.77
2016-12-05  110.00  745.00  757.71  59.70  40.02
2016-12-06  109.50  763.99  764.73  60.43  40.31
2016-12-07  109.26  764.55  761.00  60.01  39.98

[1745 rows x 5 columns]

In [19]:
# Pull the Volume dataframe
p['Volume']

AAPL        AMZN       GOOG         MSFT        YHOO
Date                                                                   
2010-01-04  123432050.0   7600543.0        NaN   38414185.0  16588957.0
2010-01-05  150476004.0   8856456.0        NaN   49758862.0  11718126.0
2010-01-06  138039594.0   7180977.0        NaN   58182332.0  16421960.0
2010-01-07  119282324.0  11030124.0        NaN   50564285.0  31816301.0
2010-01-08  111969081.0   9833829.0        NaN   51201289.0  15471074.0
2010-01-11  115557365.0   8786668.0        NaN   68754648.0  16185267.0
2010-01-12  148614774.0   9098190.0        NaN   65913228.0  15674461.0
2010-01-13  151472335.0  10727856.0        NaN   51863463.0  16961731.0
2010-01-14  108288411.0   9788435.0        NaN   63244767.0  16717061.0
2010-01-15  148584065.0  15382763.0        NaN   79915648.0  18415868.0
2010-01-19  182501620.0   8900116.0        NaN   46582886.0  15182984.0
2010-01-20  153037892.0   9081533.0        NaN   54857712.0  14421141.0
2010-01-21  152038565.0   9976146.0        NaN   73086701.0  21860505.0
2010-01-22  220441872.0  11577818.0        NaN  102004573.0  25135493.0
2010-01-25  266424802.0  12031053.0        NaN   63381173.0  19683693.0
2010-01-26  466777486.0   9567607.0        NaN   66647676.0  43979302.0
2010-01-27  430642030.0  14765273.0        NaN   63949492.0  41700911.0
2010-01-28  293375376.0  27293062.0        NaN  117513692.0  30159437.0
2010-01-29  311487785.0  29478976.0        NaN  193888424.0  39667312.0
2010-02-01  187468421.0  37774317.0        NaN   85931099.0  29865661.0
2010-02-02  174585313.0  23084986.0        NaN   54413654.0  27556624.0
2010-02-03  153831769.0  12409012.0        NaN   61397848.0  24730564.0
2010-02-04  189412762.0  12783912.0        NaN   77849968.0  27668017.0
2010-02-05  212576350.0  11027086.0        NaN   80960063.0  20713744.0
2010-02-08  119567119.0   9898391.0        NaN   52826474.0  19856359.0
2010-02-09  158221028.0   9227158.0        NaN   59195742.0  16716845.0
2010-02-10   92590218.0   6236250.0        NaN   48591246.0  36518008.0
2010-02-11  137585903.0   8344469.0        NaN   65995913.0  24510127.0
2010-02-12  163866906.0   8075533.0        NaN   81121515.0  18929239.0
2010-02-16  135987761.0   8938494.0        NaN   51938098.0  21453147.0
...                 ...         ...        ...          ...         ...
2016-10-26   66134219.0   3998102.0  1647733.0   29911608.0   6123026.0
2016-10-27   34562045.0   7406385.0  2749221.0   28479856.0   7240848.0
2016-10-28   37861662.0  10841073.0  4269902.0   33574684.0   6197469.0
2016-10-31   26419398.0   5413326.0  2427284.0   26434697.0   3749658.0
2016-11-01   43825812.0   5305428.0  2406356.0   24532986.0  11751521.0
2016-11-02   28331709.0   5026504.0  1918414.0   22147005.0  13019956.0
2016-11-03   26932602.0   3872496.0  1943175.0   21600427.0   8112820.0
2016-11-04   30836997.0   5122103.0  2134812.0   28697016.0   4579437.0
2016-11-07   32560000.0   5984354.0  1585070.0   31664798.0   5430744.0
2016-11-08   24254179.0   3412629.0  1366873.0   22935355.0   4229743.0
2016-11-09   59176361.0   8562892.0  2607121.0   49632479.0  24000065.0
2016-11-10   57134541.0  12746994.0  4745183.0   57822394.0  16443786.0
2016-11-11   34143898.0   6622784.0  2431815.0   38767843.0  15178907.0
2016-11-14   51175504.0   7321344.0  3654385.0   41328422.0  12692293.0
2016-11-15   32264510.0   6755785.0  2384001.0   35904126.0  13364918.0
2016-11-16   58840522.0   3648791.0  1472594.0   27332475.0  15609553.0
2016-11-17   26964598.0   3651345.0  1286961.0   31463728.0   8692013.0
2016-11-18   28428917.0   4373408.0  1547145.0   27686311.0   8138928.0
2016-11-21   29264571.0   4614574.0  1330639.0   19652595.0  11355856.0
2016-11-22   25965534.0   5311320.0  1593108.0   23206700.0   8881541.0
2016-11-23   27426394.0   3540263.0  1478417.0   21848913.0   3719257.0
2016-11-25   11475922.0   1837068.0   587421.0    8409616.0   2586908.0
2016-11-28   27193983.0   4438828.0  2188151.0   20732619.0

Alternate syntax is to use the dot notation, but it has the same drawbacks as other places it is availabe (namely, spaces in the items will break it

In [20]:
p.Volume

AAPL        AMZN       GOOG         MSFT        YHOO
Date                                                                   
2010-01-04  123432050.0   7600543.0        NaN   38414185.0  16588957.0
2010-01-05  150476004.0   8856456.0        NaN   49758862.0  11718126.0
2010-01-06  138039594.0   7180977.0        NaN   58182332.0  16421960.0
2010-01-07  119282324.0  11030124.0        NaN   50564285.0  31816301.0
2010-01-08  111969081.0   9833829.0        NaN   51201289.0  15471074.0
2010-01-11  115557365.0   8786668.0        NaN   68754648.0  16185267.0
2010-01-12  148614774.0   9098190.0        NaN   65913228.0  15674461.0
2010-01-13  151472335.0  10727856.0        NaN   51863463.0  16961731.0
2010-01-14  108288411.0   9788435.0        NaN   63244767.0  16717061.0
2010-01-15  148584065.0  15382763.0        NaN   79915648.0  18415868.0
2010-01-19  182501620.0   8900116.0        NaN   46582886.0  15182984.0
2010-01-20  153037892.0   9081533.0        NaN   54857712.0  14421141.0
2010-01-21  152038565.0   9976146.0        NaN   73086701.0  21860505.0
2010-01-22  220441872.0  11577818.0        NaN  102004573.0  25135493.0
2010-01-25  266424802.0  12031053.0        NaN   63381173.0  19683693.0
2010-01-26  466777486.0   9567607.0        NaN   66647676.0  43979302.0
2010-01-27  430642030.0  14765273.0        NaN   63949492.0  41700911.0
2010-01-28  293375376.0  27293062.0        NaN  117513692.0  30159437.0
2010-01-29  311487785.0  29478976.0        NaN  193888424.0  39667312.0
2010-02-01  187468421.0  37774317.0        NaN   85931099.0  29865661.0
2010-02-02  174585313.0  23084986.0        NaN   54413654.0  27556624.0
2010-02-03  153831769.0  12409012.0        NaN   61397848.0  24730564.0
2010-02-04  189412762.0  12783912.0        NaN   77849968.0  27668017.0
2010-02-05  212576350.0  11027086.0        NaN   80960063.0  20713744.0
2010-02-08  119567119.0   9898391.0        NaN   52826474.0  19856359.0
2010-02-09  158221028.0   9227158.0        NaN   59195742.0  16716845.0
2010-02-10   92590218.0   6236250.0        NaN   48591246.0  36518008.0
2010-02-11  137585903.0   8344469.0        NaN   65995913.0  24510127.0
2010-02-12  163866906.0   8075533.0        NaN   81121515.0  18929239.0
2010-02-16  135987761.0   8938494.0        NaN   51938098.0  21453147.0
...                 ...         ...        ...          ...         ...
2016-10-26   66134219.0   3998102.0  1647733.0   29911608.0   6123026.0
2016-10-27   34562045.0   7406385.0  2749221.0   28479856.0   7240848.0
2016-10-28   37861662.0  10841073.0  4269902.0   33574684.0   6197469.0
2016-10-31   26419398.0   5413326.0  2427284.0   26434697.0   3749658.0
2016-11-01   43825812.0   5305428.0  2406356.0   24532986.0  11751521.0
2016-11-02   28331709.0   5026504.0  1918414.0   22147005.0  13019956.0
2016-11-03   26932602.0   3872496.0  1943175.0   21600427.0   8112820.0
2016-11-04   30836997.0   5122103.0  2134812.0   28697016.0   4579437.0
2016-11-07   32560000.0   5984354.0  1585070.0   31664798.0   5430744.0
2016-11-08   24254179.0   3412629.0  1366873.0   22935355.0   4229743.0
2016-11-09   59176361.0   8562892.0  2607121.0   49632479.0  24000065.0
2016-11-10   57134541.0  12746994.0  4745183.0   57822394.0  16443786.0
2016-11-11   34143898.0   6622784.0  2431815.0   38767843.0  15178907.0
2016-11-14   51175504.0   7321344.0  3654385.0   41328422.0  12692293.0
2016-11-15   32264510.0   6755785.0  2384001.0   35904126.0  13364918.0
2016-11-16   58840522.0   3648791.0  1472594.0   27332475.0  15609553.0
2016-11-17   26964598.0   3651345.0  1286961.0   31463728.0   8692013.0
2016-11-18   28428917.0   4373408.0  1547145.0   27686311.0   8138928.0
2016-11-21   29264571.0   4614574.0  1330639.0   19652595.0  11355856.0
2016-11-22   25965534.0   5311320.0  1593108.0   23206700.0   8881541.0
2016-11-23   27426394.0   3540263.0  1478417.0   21848913.0   3719257.0
2016-11-25   11475922.0   1837068.0   587421.0    8409616.0   2586908.0
2016-11-28   27193983.0   4438828.0  2188151.0   20732619.0

# Section 11; Part 150
Extracting with `loc`, `iloc` and `ix` Methods

In [21]:
# Extract using index labels (loc)
# Extracting close dataframe
p.loc["Close"]

AAPL    AMZN    GOOG   MSFT   YHOO
Date                                            
2010-01-04   30.57  133.90  313.06  30.95  17.10
2010-01-05   30.63  134.69  311.68  30.96  17.23
2010-01-06   30.14  132.25  303.83  30.77  17.17
2010-01-07   30.08  130.00  296.75  30.45  16.70
2010-01-08   30.28  133.52  300.71  30.66  16.70
2010-01-11   30.02  130.31  300.25  30.27  16.74
2010-01-12   29.67  127.35  294.94  30.07  16.68
2010-01-13   30.09  129.11  293.25  30.35  16.90
2010-01-14   29.92  127.35  294.63  30.96  17.12
2010-01-15   29.42  127.14  289.71  30.86  16.82
2010-01-19   30.72  127.61  293.52  31.10  16.75
2010-01-20   30.25  125.78  289.91  30.58  16.38
2010-01-21   29.72  126.62  291.20  30.01  16.20
2010-01-22   28.25  121.43  274.73  28.96  15.88
2010-01-25   29.01  120.31  269.73  29.32  15.86
2010-01-26   29.42  119.48  270.94  29.50  15.99
2010-01-27   29.70  122.75  270.78  29.67  15.98
2010-01-28   28.47  126.03  266.88  29.16  15.44
2010-01-29   27.44  125.41  264.71  28.18  15.01
2010-02-01   27.82  118.87  266.24  28.41  15.05
2010-02-02   27.98  118.12  265.29  28.46  15.17
2010-02-03   28.46  119.10  270.14  28.63  15.46
2010-02-04   27.44  115.94  263.13  27.84  15.01
2010-02-05   27.92  117.39  265.38  28.02  15.19
2010-02-08   27.73  116.83  266.47  27.72  14.99
2010-02-09   28.03  118.03  267.95  28.01  15.07
2010-02-10   27.87  117.36  266.96  27.99  14.80
2010-02-11   28.38  120.09  267.93  28.12  15.22
2010-02-12   28.63  119.66  266.29  27.93  15.17
2010-02-16   29.06  117.53  270.38  28.35  15.41
...            ...     ...     ...    ...    ...
2016-10-26  115.59  822.59  799.07  60.63  42.08
2016-10-27  114.48  818.36  795.35  60.10  41.87
2016-10-28  113.72  776.32  795.37  59.87  41.78
2016-10-31  113.54  789.82  784.54  59.92  41.55
2016-11-01  111.49  785.41  783.61  59.80  41.33
2016-11-02  111.59  765.56  768.70  59.43  40.68
2016-11-03  109.83  767.03  762.13  59.21  40.23
2016-11-04  108.84  755.05  762.02  58.71  40.28
2016-11-07  110.41  784.93  782.52  60.42  41.05
2016-11-08  111.06  787.75  790.51  60.47  41.16
2016-11-09  110.88  771.88  785.31  60.17  41.21
2016-11-10  107.79  742.38  762.56  58.70  40.16
2016-11-11  108.43  739.01  754.02  59.02  40.42
2016-11-14  105.71  719.07  736.08  58.12  39.30
2016-11-15  107.11  743.24  758.49  58.87  40.21
2016-11-16  109.99  746.49  764.48  59.65  40.98
2016-11-17  109.95  756.40  771.23  60.64  41.45
2016-11-18  110.06  760.16  760.54  60.35  41.19
2016-11-21  111.73  780.00  769.20  60.86  41.11
2016-11-22  111.80  785.33  768.27  61.12  41.01
2016-11-23  111.23  780.12  760.99  60.40  40.96
2016-11-25  111.79  780.37  761.68  60.53  40.87
2016-11-28  111.57  766.77  768.24  60.61  41.45
2016-11-29  111.46  762.52  770.84  61.09  41.60
2016-11-30  110.52  750.57  758.04  60.26  41.02
2016-12-01  109.49  743.65  747.92  59.20  39.63
2016-12-02  109.90  740.34  750.50  59.25  40.07
2016-12-05  109.11  759.36  762.52  60.22  40.20
2016-12-06  109.95  764.72  759.11  59.95  39.97
2016-12-07  111.03  770.42  771.19  61.37  40.52

[1745 rows x 5 columns]

In [22]:
# Get Close values for a specific date
p.loc["Close", '2010-01-04']

AAPL     30.57
AMZN    133.90
GOOG    313.06
MSFT     30.95
YHOO     17.10
Name: 2010-01-04 00:00:00, dtype: float64

In [23]:
# Get close value for MSFT on 2010-01-04
p.loc["Close", '2010-01-04', "MSFT"]

30.949999999999999

In [24]:
# Extract using index position (iloc)
p.items

Index([u'Open', u'High', u'Low', u'Close', u'Volume'], dtype='object')

In [25]:
# Get close (3rd index)
p.iloc[3]

AAPL    AMZN    GOOG   MSFT   YHOO
Date                                            
2010-01-04   30.57  133.90  313.06  30.95  17.10
2010-01-05   30.63  134.69  311.68  30.96  17.23
2010-01-06   30.14  132.25  303.83  30.77  17.17
2010-01-07   30.08  130.00  296.75  30.45  16.70
2010-01-08   30.28  133.52  300.71  30.66  16.70
2010-01-11   30.02  130.31  300.25  30.27  16.74
2010-01-12   29.67  127.35  294.94  30.07  16.68
2010-01-13   30.09  129.11  293.25  30.35  16.90
2010-01-14   29.92  127.35  294.63  30.96  17.12
2010-01-15   29.42  127.14  289.71  30.86  16.82
2010-01-19   30.72  127.61  293.52  31.10  16.75
2010-01-20   30.25  125.78  289.91  30.58  16.38
2010-01-21   29.72  126.62  291.20  30.01  16.20
2010-01-22   28.25  121.43  274.73  28.96  15.88
2010-01-25   29.01  120.31  269.73  29.32  15.86
2010-01-26   29.42  119.48  270.94  29.50  15.99
2010-01-27   29.70  122.75  270.78  29.67  15.98
2010-01-28   28.47  126.03  266.88  29.16  15.44
2010-01-29   27.44  125.41  264.71  28.18  15.01
2010-02-01   27.82  118.87  266.24  28.41  15.05
2010-02-02   27.98  118.12  265.29  28.46  15.17
2010-02-03   28.46  119.10  270.14  28.63  15.46
2010-02-04   27.44  115.94  263.13  27.84  15.01
2010-02-05   27.92  117.39  265.38  28.02  15.19
2010-02-08   27.73  116.83  266.47  27.72  14.99
2010-02-09   28.03  118.03  267.95  28.01  15.07
2010-02-10   27.87  117.36  266.96  27.99  14.80
2010-02-11   28.38  120.09  267.93  28.12  15.22
2010-02-12   28.63  119.66  266.29  27.93  15.17
2010-02-16   29.06  117.53  270.38  28.35  15.41
...            ...     ...     ...    ...    ...
2016-10-26  115.59  822.59  799.07  60.63  42.08
2016-10-27  114.48  818.36  795.35  60.10  41.87
2016-10-28  113.72  776.32  795.37  59.87  41.78
2016-10-31  113.54  789.82  784.54  59.92  41.55
2016-11-01  111.49  785.41  783.61  59.80  41.33
2016-11-02  111.59  765.56  768.70  59.43  40.68
2016-11-03  109.83  767.03  762.13  59.21  40.23
2016-11-04  108.84  755.05  762.02  58.71  40.28
2016-11-07  110.41  784.93  782.52  60.42  41.05
2016-11-08  111.06  787.75  790.51  60.47  41.16
2016-11-09  110.88  771.88  785.31  60.17  41.21
2016-11-10  107.79  742.38  762.56  58.70  40.16
2016-11-11  108.43  739.01  754.02  59.02  40.42
2016-11-14  105.71  719.07  736.08  58.12  39.30
2016-11-15  107.11  743.24  758.49  58.87  40.21
2016-11-16  109.99  746.49  764.48  59.65  40.98
2016-11-17  109.95  756.40  771.23  60.64  41.45
2016-11-18  110.06  760.16  760.54  60.35  41.19
2016-11-21  111.73  780.00  769.20  60.86  41.11
2016-11-22  111.80  785.33  768.27  61.12  41.01
2016-11-23  111.23  780.12  760.99  60.40  40.96
2016-11-25  111.79  780.37  761.68  60.53  40.87
2016-11-28  111.57  766.77  768.24  60.61  41.45
2016-11-29  111.46  762.52  770.84  61.09  41.60
2016-11-30  110.52  750.57  758.04  60.26  41.02
2016-12-01  109.49  743.65  747.92  59.20  39.63
2016-12-02  109.90  740.34  750.50  59.25  40.07
2016-12-05  109.11  759.36  762.52  60.22  40.20
2016-12-06  109.95  764.72  759.11  59.95  39.97
2016-12-07  111.03  770.42  771.19  61.37  40.52

[1745 rows x 5 columns]

In [26]:
# Get 2010-01-04 (0) in close dataframe 
p.iloc[3, 0]

AAPL     30.57
AMZN    133.90
GOOG    313.06
MSFT     30.95
YHOO     17.10
Name: 2010-01-04 00:00:00, dtype: float64

In [27]:
# Get MSFT (3rd index) close value on date above 
p.iloc[3, 0, 3]

30.949999999999999

In [28]:
# Mix/match position and labels using .ix
p.ix["Close", 0, "GOOG"]

313.06

# Section 11; Part 151
Convert panel into a multiindex dataframe (and reverse)

In [31]:
# Convert panel to multi-index dataframe
df = p.to_frame()
df.head()

Open    High     Low   Close       Volume
Date       minor                                             
2010-01-04 AAPL    30.49   30.64   30.34   30.57  123432050.0
           AMZN   136.25  136.61  133.14  133.90    7600543.0
           MSFT    30.62   31.10   30.59   30.95   38414185.0
           YHOO    16.94   17.20   16.88   17.10   16588957.0
2010-01-05 AAPL    30.66   30.80   30.46   30.63  150476004.0

 - Items axis became column headers
 - Major axis became outer index
 - Minor axis became inner index

In [33]:
# Convert multi-index dataframe to a panel
p2 = df.to_panel()
p2

<class 'pandas.core.panel.Panel'>
Dimensions: 5 (items) x 1745 (major_axis) x 5 (minor_axis)
Items axis: Open to Volume
Major_axis axis: 2010-01-04 00:00:00 to 2016-12-07 00:00:00
Minor_axis axis: AAPL to YHOO

# Section 11; Part 152
The `major_xs()` method

 - Will remove the major_axis and return a dataframe for that specific major axis value

In [34]:
# Have the major_axis *attribute* - Returns values that are the row labels in the panel
p.major_axis

DatetimeIndex(['2010-01-04', '2010-01-05', '2010-01-06', '2010-01-07',
               '2010-01-08', '2010-01-11', '2010-01-12', '2010-01-13',
               '2010-01-14', '2010-01-15',
               ...
               '2016-11-23', '2016-11-25', '2016-11-28', '2016-11-29',
               '2016-11-30', '2016-12-01', '2016-12-02', '2016-12-05',
               '2016-12-06', '2016-12-07'],
              dtype='datetime64[ns]', name=u'Date', length=1745, freq=None)

In [36]:
# major_xs() is a *method* - Returns a dataframe where the minor access is the Index and the Items axis are the columns

#  Get data for 2010-01-04
p.major_xs("2010-01-04")

Open    High     Low   Close       Volume
AAPL   30.49   30.64   30.34   30.57  123432050.0
AMZN  136.25  136.61  133.14  133.90    7600543.0
GOOG  313.16  314.44  311.81  313.06          NaN
MSFT   30.62   31.10   30.59   30.95   38414185.0
YHOO   16.94   17.20   16.88   17.10   16588957.0

# Section 11; Part 153
The `minor_xs()` method

 - Remove the minor_axis and return a dataframe for that specific minor axis value

In [37]:
p.minor_axis

Index([u'AAPL', u'AMZN', u'GOOG', u'MSFT', u'YHOO'], dtype='object')

In [39]:
# Get all of AAPL's stock values
p.minor_xs('AAPL')

Open    High     Low   Close       Volume
Date                                                   
2010-01-04   30.49   30.64   30.34   30.57  123432050.0
2010-01-05   30.66   30.80   30.46   30.63  150476004.0
2010-01-06   30.63   30.75   30.11   30.14  138039594.0
2010-01-07   30.25   30.29   29.86   30.08  119282324.0
2010-01-08   30.04   30.29   29.87   30.28  111969081.0
2010-01-11   30.40   30.43   29.78   30.02  115557365.0
2010-01-12   29.88   29.97   29.49   29.67  148614774.0
2010-01-13   29.70   30.13   29.16   30.09  151472335.0
2010-01-14   30.02   30.07   29.86   29.92  108288411.0
2010-01-15   30.13   30.23   29.41   29.42  148584065.0
2010-01-19   29.76   30.74   29.61   30.72  182501620.0
2010-01-20   30.70   30.79   29.93   30.25  153037892.0
2010-01-21   30.30   30.47   29.60   29.72  152038565.0
2010-01-22   29.54   29.64   28.17   28.25  220441872.0
2010-01-25   28.93   29.24   28.60   29.01  266424802.0
2010-01-26   29.42   30.53   28.94   29.42  466777486.0
2010-01-27   29.55   30.08   28.50   29.70  430642030.0
2010-01-28   29.28   29.36   28.39   28.47  293375376.0
2010-01-29   28.73   28.89   27.18   27.44  311487785.0
2010-02-01   27.48   28.00   27.33   27.82  187468421.0
2010-02-02   27.99   28.05   27.63   27.98  174585313.0
2010-02-03   27.88   28.60   27.77   28.46  153831769.0
2010-02-04   28.10   28.34   27.37   27.44  189412762.0
2010-02-05   27.52   28.00   27.26   27.92  212576350.0
2010-02-08   27.96   28.27   27.71   27.73  119567119.0
2010-02-09   28.06   28.21   27.82   28.03  158221028.0
2010-02-10   27.98   28.09   27.75   27.87   92590218.0
2010-02-11   27.84   28.54   27.72   28.38  137585903.0
2010-02-12   28.30   28.81   27.93   28.63  163866906.0
2010-02-16   28.85   29.10   28.79   29.06  135987761.0
...            ...     ...     ...     ...          ...
2016-10-26  114.31  115.70  113.31  115.59   66134219.0
2016-10-27  115.39  115.86  114.10  114.48   34562045.0
2016-10-28  113.87  115.21  113.45  113.72   37861662.0
2016-10-31  113.65  114.23  113.20  113.54   26419398.0
2016-11-01  113.46  113.77  110.53  111.49   43825812.0
2016-11-02  111.40  112.35  111.23  111.59   28331709.0
2016-11-03  110.98  111.46  109.55  109.83   26932602.0
2016-11-04  108.53  110.25  108.11  108.84   30836997.0
2016-11-07  110.08  110.51  109.46  110.41   32560000.0
2016-11-08  110.31  111.72  109.70  111.06   24254179.0
2016-11-09  109.88  111.32  108.05  110.88   59176361.0
2016-11-10  111.09  111.09  105.83  107.79   57134541.0
2016-11-11  107.12  108.87  106.55  108.43   34143898.0
2016-11-14  107.71  107.81  104.08  105.71   51175504.0
2016-11-15  106.57  107.68  106.16  107.11   32264510.0
2016-11-16  106.70  110.23  106.60  109.99   58840522.0
2016-11-17  109.81  110.35  108.83  109.95   26964598.0
2016-11-18  109.72  110.54  109.66  110.06   28428917.0
2016-11-21  110.12  111.99  110.01  111.73   29264571.0
2016-11-22  111.95  112.42  111.40  111.80   25965534.0
2016-11-23  111.36  111.51  110.33  111.23   27426394.0
2016-11-25  111.47  111.87  110.95  111.79   11475922.0
2016-11-28  111.43  112.46  111.39  111.57   27193983.0
2016-11-29  110.78  112.03  110.07  111.46   28528750.0
2016-11-30  111.60  112.20  110.27  110.52   36162258.0
2016-12-01  110.36  110.94  109.03  109.49   37086862.0
2016-12-02  109.17  110.09  108.85  109.90   26527997.0
2016-12-05  110.00  110.03  108.25  109.11   34324540.0
2016-12-06  109.50  110.36  109.19  109.95   26195462.0
2016-12-07  109.26  111.19  109.16  111.03   29998719.0

[1745 rows x 5 columns]

# Section 11; Part 154
Transpose a panel with `.transpose()`

 - Swap the axes that make up a panel object

In [40]:
p.axes

[Index([u'Open', u'High', u'Low', u'Close', u'Volume'], dtype='object'),
 DatetimeIndex(['2010-01-04', '2010-01-05', '2010-01-06', '2010-01-07',
                '2010-01-08', '2010-01-11', '2010-01-12', '2010-01-13',
                '2010-01-14', '2010-01-15',
                ...
                '2016-11-23', '2016-11-25', '2016-11-28', '2016-11-29',
                '2016-11-30', '2016-12-01', '2016-12-02', '2016-12-05',
                '2016-12-06', '2016-12-07'],
               dtype='datetime64[ns]', name=u'Date', length=1745, freq=None),
 Index([u'AAPL', u'AMZN', u'GOOG', u'MSFT', u'YHOO'], dtype='object')]

In [42]:
# Swap Stock information (Open/Close/High/Low/Volume) with the Company information
#   This is the items axis and the minor axis
# Feed it a list of indexes the new panel will have in the order we want 

# Items - Index position of 0
# Major - Index position of 1
# Minor - Index position of 2
p2 = p.transpose(2, 1 ,0)
p2

<class 'pandas.core.panel.Panel'>
Dimensions: 5 (items) x 1745 (major_axis) x 5 (minor_axis)
Items axis: AAPL to YHOO
Major_axis axis: 2010-01-04 00:00:00 to 2016-12-07 00:00:00
Minor_axis axis: Open to Volume

In [43]:
# Now can get company information like this, instead of using the .minor_xs() method
p2['AAPL']

Open    High     Low   Close       Volume
Date                                                   
2010-01-04   30.49   30.64   30.34   30.57  123432050.0
2010-01-05   30.66   30.80   30.46   30.63  150476004.0
2010-01-06   30.63   30.75   30.11   30.14  138039594.0
2010-01-07   30.25   30.29   29.86   30.08  119282324.0
2010-01-08   30.04   30.29   29.87   30.28  111969081.0
2010-01-11   30.40   30.43   29.78   30.02  115557365.0
2010-01-12   29.88   29.97   29.49   29.67  148614774.0
2010-01-13   29.70   30.13   29.16   30.09  151472335.0
2010-01-14   30.02   30.07   29.86   29.92  108288411.0
2010-01-15   30.13   30.23   29.41   29.42  148584065.0
2010-01-19   29.76   30.74   29.61   30.72  182501620.0
2010-01-20   30.70   30.79   29.93   30.25  153037892.0
2010-01-21   30.30   30.47   29.60   29.72  152038565.0
2010-01-22   29.54   29.64   28.17   28.25  220441872.0
2010-01-25   28.93   29.24   28.60   29.01  266424802.0
2010-01-26   29.42   30.53   28.94   29.42  466777486.0
2010-01-27   29.55   30.08   28.50   29.70  430642030.0
2010-01-28   29.28   29.36   28.39   28.47  293375376.0
2010-01-29   28.73   28.89   27.18   27.44  311487785.0
2010-02-01   27.48   28.00   27.33   27.82  187468421.0
2010-02-02   27.99   28.05   27.63   27.98  174585313.0
2010-02-03   27.88   28.60   27.77   28.46  153831769.0
2010-02-04   28.10   28.34   27.37   27.44  189412762.0
2010-02-05   27.52   28.00   27.26   27.92  212576350.0
2010-02-08   27.96   28.27   27.71   27.73  119567119.0
2010-02-09   28.06   28.21   27.82   28.03  158221028.0
2010-02-10   27.98   28.09   27.75   27.87   92590218.0
2010-02-11   27.84   28.54   27.72   28.38  137585903.0
2010-02-12   28.30   28.81   27.93   28.63  163866906.0
2010-02-16   28.85   29.10   28.79   29.06  135987761.0
...            ...     ...     ...     ...          ...
2016-10-26  114.31  115.70  113.31  115.59   66134219.0
2016-10-27  115.39  115.86  114.10  114.48   34562045.0
2016-10-28  113.87  115.21  113.45  113.72   37861662.0
2016-10-31  113.65  114.23  113.20  113.54   26419398.0
2016-11-01  113.46  113.77  110.53  111.49   43825812.0
2016-11-02  111.40  112.35  111.23  111.59   28331709.0
2016-11-03  110.98  111.46  109.55  109.83   26932602.0
2016-11-04  108.53  110.25  108.11  108.84   30836997.0
2016-11-07  110.08  110.51  109.46  110.41   32560000.0
2016-11-08  110.31  111.72  109.70  111.06   24254179.0
2016-11-09  109.88  111.32  108.05  110.88   59176361.0
2016-11-10  111.09  111.09  105.83  107.79   57134541.0
2016-11-11  107.12  108.87  106.55  108.43   34143898.0
2016-11-14  107.71  107.81  104.08  105.71   51175504.0
2016-11-15  106.57  107.68  106.16  107.11   32264510.0
2016-11-16  106.70  110.23  106.60  109.99   58840522.0
2016-11-17  109.81  110.35  108.83  109.95   26964598.0
2016-11-18  109.72  110.54  109.66  110.06   28428917.0
2016-11-21  110.12  111.99  110.01  111.73   29264571.0
2016-11-22  111.95  112.42  111.40  111.80   25965534.0
2016-11-23  111.36  111.51  110.33  111.23   27426394.0
2016-11-25  111.47  111.87  110.95  111.79   11475922.0
2016-11-28  111.43  112.46  111.39  111.57   27193983.0
2016-11-29  110.78  112.03  110.07  111.46   28528750.0
2016-11-30  111.60  112.20  110.27  110.52   36162258.0
2016-12-01  110.36  110.94  109.03  109.49   37086862.0
2016-12-02  109.17  110.09  108.85  109.90   26527997.0
2016-12-05  110.00  110.03  108.25  109.11   34324540.0
2016-12-06  109.50  110.36  109.19  109.95   26195462.0
2016-12-07  109.26  111.19  109.16  111.03   29998719.0

[1745 rows x 5 columns]

In [44]:
# Using major_xs() and minor_xs() still works
# Get information for all on a specific day
p2.major_xs('2010-01-04')

AAPL        AMZN    GOOG         MSFT         YHOO
Open    3.049000e+01      136.25  313.16        30.62        16.94
High    3.064000e+01      136.61  314.44        31.10        17.20
Low     3.034000e+01      133.14  311.81        30.59        16.88
Close   3.057000e+01      133.90  313.06        30.95        17.10
Volume  1.234320e+08  7600543.00     NaN  38414185.00  16588957.00

In [45]:
# Get all volumes
p2.minor_xs("Volume")

AAPL        AMZN       GOOG         MSFT        YHOO
Date                                                                   
2010-01-04  123432050.0   7600543.0        NaN   38414185.0  16588957.0
2010-01-05  150476004.0   8856456.0        NaN   49758862.0  11718126.0
2010-01-06  138039594.0   7180977.0        NaN   58182332.0  16421960.0
2010-01-07  119282324.0  11030124.0        NaN   50564285.0  31816301.0
2010-01-08  111969081.0   9833829.0        NaN   51201289.0  15471074.0
2010-01-11  115557365.0   8786668.0        NaN   68754648.0  16185267.0
2010-01-12  148614774.0   9098190.0        NaN   65913228.0  15674461.0
2010-01-13  151472335.0  10727856.0        NaN   51863463.0  16961731.0
2010-01-14  108288411.0   9788435.0        NaN   63244767.0  16717061.0
2010-01-15  148584065.0  15382763.0        NaN   79915648.0  18415868.0
2010-01-19  182501620.0   8900116.0        NaN   46582886.0  15182984.0
2010-01-20  153037892.0   9081533.0        NaN   54857712.0  14421141.0
2010-01-21  152038565.0   9976146.0        NaN   73086701.0  21860505.0
2010-01-22  220441872.0  11577818.0        NaN  102004573.0  25135493.0
2010-01-25  266424802.0  12031053.0        NaN   63381173.0  19683693.0
2010-01-26  466777486.0   9567607.0        NaN   66647676.0  43979302.0
2010-01-27  430642030.0  14765273.0        NaN   63949492.0  41700911.0
2010-01-28  293375376.0  27293062.0        NaN  117513692.0  30159437.0
2010-01-29  311487785.0  29478976.0        NaN  193888424.0  39667312.0
2010-02-01  187468421.0  37774317.0        NaN   85931099.0  29865661.0
2010-02-02  174585313.0  23084986.0        NaN   54413654.0  27556624.0
2010-02-03  153831769.0  12409012.0        NaN   61397848.0  24730564.0
2010-02-04  189412762.0  12783912.0        NaN   77849968.0  27668017.0
2010-02-05  212576350.0  11027086.0        NaN   80960063.0  20713744.0
2010-02-08  119567119.0   9898391.0        NaN   52826474.0  19856359.0
2010-02-09  158221028.0   9227158.0        NaN   59195742.0  16716845.0
2010-02-10   92590218.0   6236250.0        NaN   48591246.0  36518008.0
2010-02-11  137585903.0   8344469.0        NaN   65995913.0  24510127.0
2010-02-12  163866906.0   8075533.0        NaN   81121515.0  18929239.0
2010-02-16  135987761.0   8938494.0        NaN   51938098.0  21453147.0
...                 ...         ...        ...          ...         ...
2016-10-26   66134219.0   3998102.0  1647733.0   29911608.0   6123026.0
2016-10-27   34562045.0   7406385.0  2749221.0   28479856.0   7240848.0
2016-10-28   37861662.0  10841073.0  4269902.0   33574684.0   6197469.0
2016-10-31   26419398.0   5413326.0  2427284.0   26434697.0   3749658.0
2016-11-01   43825812.0   5305428.0  2406356.0   24532986.0  11751521.0
2016-11-02   28331709.0   5026504.0  1918414.0   22147005.0  13019956.0
2016-11-03   26932602.0   3872496.0  1943175.0   21600427.0   8112820.0
2016-11-04   30836997.0   5122103.0  2134812.0   28697016.0   4579437.0
2016-11-07   32560000.0   5984354.0  1585070.0   31664798.0   5430744.0
2016-11-08   24254179.0   3412629.0  1366873.0   22935355.0   4229743.0
2016-11-09   59176361.0   8562892.0  2607121.0   49632479.0  24000065.0
2016-11-10   57134541.0  12746994.0  4745183.0   57822394.0  16443786.0
2016-11-11   34143898.0   6622784.0  2431815.0   38767843.0  15178907.0
2016-11-14   51175504.0   7321344.0  3654385.0   41328422.0  12692293.0
2016-11-15   32264510.0   6755785.0  2384001.0   35904126.0  13364918.0
2016-11-16   58840522.0   3648791.0  1472594.0   27332475.0  15609553.0
2016-11-17   26964598.0   3651345.0  1286961.0   31463728.0   8692013.0
2016-11-18   28428917.0   4373408.0  1547145.0   27686311.0   8138928.0
2016-11-21   29264571.0   4614574.0  1330639.0   19652595.0  11355856.0
2016-11-22   25965534.0   5311320.0  1593108.0   23206700.0   8881541.0
2016-11-23   27426394.0   3540263.0  1478417.0   21848913.0   3719257.0
2016-11-25   11475922.0   1837068.0   587421.0    8409616.0   2586908.0
2016-11-28   27193983.0   4438828.0  2188151.0   20732619.0

# Section 11; Part 155
The `swapaxes()` Method

 - Alternative to .transpose() without need for knowing index positions of each axes

In [47]:
# Swap Stock information (Open/Close/High/Low/Volume) with the Company information
p3 = p.swapaxes("items", "minor")
p3

<class 'pandas.core.panel.Panel'>
Dimensions: 5 (items) x 1745 (major_axis) x 5 (minor_axis)
Items axis: AAPL to YHOO
Major_axis axis: 2010-01-04 00:00:00 to 2016-12-07 00:00:00
Minor_axis axis: Open to Volume

In [48]:
p3['MSFT']

Open   High    Low  Close       Volume
Date                                               
2010-01-04  30.62  31.10  30.59  30.95   38414185.0
2010-01-05  30.85  31.10  30.64  30.96   49758862.0
2010-01-06  30.88  31.08  30.52  30.77   58182332.0
2010-01-07  30.63  30.70  30.19  30.45   50564285.0
2010-01-08  30.28  30.88  30.24  30.66   51201289.0
2010-01-11  30.71  30.76  30.12  30.27   68754648.0
2010-01-12  30.15  30.40  29.91  30.07   65913228.0
2010-01-13  30.26  30.52  30.01  30.35   51863463.0
2010-01-14  30.31  31.10  30.26  30.96   63244767.0
2010-01-15  31.08  31.24  30.71  30.86   79915648.0
2010-01-19  30.75  31.24  30.68  31.10   46582886.0
2010-01-20  30.81  30.94  30.31  30.58   54857712.0
2010-01-21  30.61  30.72  30.00  30.01   73086701.0
2010-01-22  30.00  30.20  28.84  28.96  102004573.0
2010-01-25  29.24  29.66  29.10  29.32   63381173.0
2010-01-26  29.20  29.85  29.09  29.50   66647676.0
2010-01-27  29.35  29.82  29.02  29.67   63949492.0
2010-01-28  29.84  29.87  28.89  29.16  117513692.0
2010-01-29  29.90  29.92  27.66  28.18  193888424.0
2010-02-01  28.39  28.48  27.92  28.41   85931099.0
2010-02-02  28.37  28.50  28.14  28.46   54413654.0
2010-02-03  28.26  28.79  28.12  28.63   61397848.0
2010-02-04  28.38  28.50  27.81  27.84   77849968.0
2010-02-05  28.00  28.28  27.57  28.02   80960063.0
2010-02-08  28.01  28.08  27.57  27.72   52826474.0
2010-02-09  27.97  28.34  27.75  28.01   59195742.0
2010-02-10  28.03  28.24  27.84  27.99   48591246.0
2010-02-11  27.93  28.40  27.70  28.12   65995913.0
2010-02-12  27.81  28.06  27.58  27.93   81121515.0
2010-02-16  28.13  28.37  28.02  28.35   51938098.0
...           ...    ...    ...    ...          ...
2016-10-26  60.81  61.20  60.47  60.63   29911608.0
2016-10-27  60.61  60.83  60.09  60.10   28479856.0
2016-10-28  60.01  60.52  59.58  59.87   33574684.0
2016-10-31  60.16  60.42  59.92  59.92   26434697.0
2016-11-01  59.97  60.02  59.25  59.80   24532986.0
2016-11-02  59.82  59.93  59.30  59.43   22147005.0
2016-11-03  59.53  59.64  59.11  59.21   21600427.0
2016-11-04  58.65  59.28  58.52  58.71   28697016.0
2016-11-07  59.78  60.52  59.78  60.42   31664798.0
2016-11-08  60.55  60.78  60.15  60.47   22935355.0
2016-11-09  60.00  60.59  59.20  60.17   49632479.0
2016-11-10  60.48  60.49  57.63  58.70   57822394.0
2016-11-11  58.23  59.12  58.01  59.02   38767843.0
2016-11-14  59.02  59.08  57.28  58.12   41328422.0
2016-11-15  58.33  59.49  58.32  58.87   35904126.0
2016-11-16  58.94  59.66  58.81  59.65   27332475.0
2016-11-17  60.41  60.95  59.96  60.64   31463728.0
2016-11-18  60.78  61.14  60.30  60.35   27686311.0
2016-11-21  60.50  60.97  60.42  60.86   19652595.0
2016-11-22  60.98  61.26  60.80  61.12   23206700.0
2016-11-23  61.01  61.10  60.25  60.40   21848913.0
2016-11-25  60.30  60.53  60.13  60.53    8409616.0
2016-11-28  60.34  61.02  60.21  60.61   20732619.0
2016-11-29  60.65  61.41  60.52  61.09   22366721.0
2016-11-30  60.86  61.18  60.22  60.26   34655435.0
2016-12-01  60.11  60.15  58.94  59.20   34542121.0
2016-12-02  59.08  59.47  58.80  59.25   25515665.0
2016-12-05  59.70  60.58  59.56  60.22   23552658.0
2016-12-06  60.43  60.46  59.80  59.95   19907035.0
2016-12-07  60.01  61.38  59.80  61.37   30808969.0

[1745 rows x 5 columns]